### PR TITLE
Feature/download status #48

### LINKED ISSUE #48: [datagateway-download] Create "download status" view
This is a large issue that may get split up into smaller issues.

Recreate the TopCAT download status view. Unlike TopCAT which has it as a modal, we should instead have it as a separate page accessed via the menu (Data/Download). I think we shall need two "tabs" - one for the download cart and one for the download status page.

Download status page mockup (based on current TopCAT):
downloads

Acceptance criteria:
- [x] Have a download status view that displays the carts the user has requested for download, and their statuses
- [x] The user can download individual carts
- [x] The user can delete individual carts

### PR DESCRIPTION
**Pull request by GoelBiju**
## Description
A download status table that shows all the download requests submitted by a user. This is now as a tab alongside the download cart and the download confirmation dialog.

## Testing instructions

- [ ] Review code
- [ ] Check Travis build
- [ ] Review changes to test coverage

## Agile board tracking
Closes #48

### CODE REVIEW

**Comment by agbeltran** [comment]
As discussed in the mid-sprint review meeting, we want to add information on a timestamp of when the download status was ```lastChecked```.

File: packages/datagateway-common/src/app.types.tsx [hunk 1]
```diff
@@ -199,14 +199,18 @@ export type DownloadCartTableItem = DownloadCartItem & {
   [key: string]: string | number | DownloadCartItem[];
 };
 
+// export type DownloadStatusTableItem = Download & {
```
Comment (louise-davies) [inline packages/datagateway-common/src/app.types.tsx]: Is this needed?

File: packages/datagateway-common/src/app.types.tsx [hunk 2]
```diff
@@ -199,14 +199,18 @@ export type DownloadCartTableItem = DownloadCartItem & {
...
+
 export type ICATEntity =
   | Investigation
   | Dataset
   | Datafile
   | Instrument
   | FacilityCycle;
 
-export type Entity = ICATEntity | DownloadCartTableItem;
+export type Entity = ICATEntity | DownloadCartTableItem | Download; // | DownloadStatusTableItem;
```
Comment (louise-davies) [inline packages/datagateway-common/src/app.types.tsx]: Again, is the comment needed?

File: packages/datagateway-download/cypress/support/commands.js [hunk 3]
```diff
@@ -57,14 +92,25 @@ export const readSciGatewayToken = () => {
   };
 };
 
-Cypress.Commands.add('login', (username, password) => {
-  cy.request('POST', `http://scigateway-preprod.esc.rl.ac.uk:5000/sessions`, {
-    username: username,
-    password: password,
+Cypress.Commands.add('login', (username, password, noRootCredentials) => {
+  let credentials = !noRootCredentials
```
Comment (louise-davies) [inline packages/datagateway-download/cypress/support/commands.js]: You can probably just change this command to accept `username` and `password` params and change any root logins to use call it with `cy.login("root", "pw")`

File: packages/datagateway-download/src/downloadConfirmation/downloadConfirmDialog.component.tsx [hunk 4]
```diff
@@ -109,16 +105,15 @@ interface DownloadConfirmDialogProps
   isTwoLevel: boolean;
   open: boolean;
 
-  // TODO: pass in the function to call to redirect to the status tab.
-  // setStatus: () => void;
+  setStatus: () => void;
```
Comment (louise-davies) [inline packages/datagateway-download/src/downloadConfirmation/downloadConfirmDialog.component.tsx]: This `setStatus` is a pretty confusing name - it took me reading the removed comments to realise what it is doing. Perhaps something like `redirectToStatusTab`?

File: packages/datagateway-download/src/downloadCart/downloadCartTable.component.tsx [hunk 5]
```diff
@@ -17,12 +17,18 @@ import {
...
   getIsTwoLevel,
-} from './downloadCartApi';
+} from '../downloadApi';
 import chunk from 'lodash.chunk';
 
 import DownloadConfirmDialog from '../downloadConfirmation/downloadConfirmDialog.component';
 
-const DownloadCartTable: React.FC = () => {
+interface DownloadCartTableProps {
+  statusLink: () => void;
```
Comment (louise-davies) [inline packages/datagateway-download/src/downloadCart/downloadCartTable.component.tsx]: Again, kind of a confusing name. `statusTabRedirect`?

File: packages/datagateway-download/src/downloadStatus/downloadStatusTable.component.tsx [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+  setLastChecked: () => void;
+}
+
+const DownloadStatusTable: React.FC<DownloadStatusTableProps> = (
+  props: DownloadStatusTableProps
+) => {
+  // Sorting columns
+  const [sort, setSort] = React.useState<{ [column: string]: Order }>({});
+  const [filters, setFilters] = React.useState<{
+    // [column: string]: string;
```
Comment (louise-davies) [inline packages/datagateway-download/src/downloadStatus/downloadStatusTable.component.tsx]: Remove comment

File: packages/datagateway-download/src/index.tsx [hunk 7]
```diff
@@ -47,6 +47,12 @@ document.addEventListener(MicroFrontendMessageId, e => {
   // attempt to re-render the plugin if the corresponding div is present
   const action = (e as CustomEvent).detail;
   if (action.type === RequestPluginRerenderType) {
+    // This is a temporary fix for the current issue with the tab indicator
+    // not updating after the size of the page has been altered.
+    setTimeout(() => {
+      window.dispatchEvent(new CustomEvent('resize'));
+      // console.log('resize done');
```
Comment (louise-davies) [inline packages/datagateway-download/src/index.tsx]: remove console.log comment

File: packages/datagateway-download/src/index.tsx [hunk 8]
```diff
@@ -47,6 +47,12 @@ document.addEventListener(MicroFrontendMessageId, e => {
   // attempt to re-render the plugin if the corresponding div is present
   const action = (e as CustomEvent).detail;
   if (action.type === RequestPluginRerenderType) {
+    // This is a temporary fix for the current issue with the tab indicator
+    // not updating after the size of the page has been altered.
```
Comment (louise-davies) [inline packages/datagateway-download/src/index.tsx]: perhaps include a link to the MUI issue in the comment?

File: packages/datagateway-download/package.json [hunk 9]
```diff
@@ -17,6 +17,7 @@
     "react": "^16.11.0",
     "react-dom": "^16.11.0",
     "react-scripts": "3.0.1",
+    "react-virtualized": "^9.21.2",
```
Comment (louise-davies) [inline packages/datagateway-download/package.json]: Can we bump this down to `^9.21.1` to match the version in `common`?

**Pull request merged**